pkg/policy/directory/watcher.go

```diff
@@ -0,0 +1,217 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package directory
+
+import (
+	"context"
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"unicode/utf8"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/sirupsen/logrus"
+	"sigs.k8s.io/yaml"
+
+	ipcacheTypes "github.com/cilium/cilium/pkg/ipcache/types"
+	k8sConst "github.com/cilium/cilium/pkg/k8s/apis/cilium.io"
+	k8sCiliumUtils "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/utils"
+	cilium_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/source"
+)
+
+type policyWatcher struct {
+	log           logrus.FieldLogger
+	config        *option.DaemonConfig
+	policyManager PolicyManager
+	readStatus    DirectoryWatcherReadStatus
+	// maps cnp file name to cnp object. this is required to retrieve data during delete.
+	fileNameToCnpCache map[string]*cilium_v2.CiliumNetworkPolicy
+}
+
+func (p *policyWatcher) translateToCNPObject(file string) (*cilium_v2.CiliumNetworkPolicy, error) {
+	data, err := os.ReadFile(file)
+	if err != nil {
+		return nil, err
+	}
+
+	// yaml to json conversion
+	jsonData, err := yaml.YAMLToJSON([]byte(data))
+	if err != nil {
+		return nil, err
+	}
+
+	// translate json to cnp object
+	cnp := &cilium_v2.CiliumNetworkPolicy{}
+	err = json.Unmarshal(jsonData, cnp)
+	if err != nil {
+		return nil, err
+	}
+
+	return cnp, nil
+}
+
+func getLabels(fileName string, cnp *cilium_v2.CiliumNetworkPolicy) labels.LabelArray {
+	labelsArr := labels.LabelArray{
+		labels.NewLabel("filename", fileName, labels.LabelSourceDirectory),
+	}
+
+	ns := cnp.ObjectMeta.Namespace
+	// For clusterwide policy namespace will be empty.
+	if ns != "" {
+		nsLabel := labels.NewLabel(k8sConst.PolicyLabelNamespace, ns, labels.LabelSourceDirectory)
+		labelsArr = append(labelsArr, nsLabel)
+		srcLabel := labels.NewLabel("policy-derived-from", k8sCiliumUtils.ResourceTypeCiliumNetworkPolicy, labels.LabelSourceDirectory)
+		labelsArr = append(labelsArr, srcLabel)
+	} else {
+		srcLabel := labels.NewLabel("policy-derived-from", k8sCiliumUtils.ResourceTypeCiliumClusterwideNetworkPolicy, labels.LabelSourceDirectory)
+		labelsArr = append(labelsArr, srcLabel)
+	}
+
+	return labelsArr
+}
+
+// read cilium network policy yaml file and convert to policy object and
+// add rules to policy engine.
+func (p *policyWatcher) addToPolicyEngine(cnpFilePath string) error {
+	// read from file and convert to cnp object
+	cnp, err := p.translateToCNPObject(cnpFilePath)
+	if err != nil {
+		return err
+	}
+
+	cnp.ObjectMeta.Name = filepath.Base(cnpFilePath)
+
+	resourceID := ipcacheTypes.NewResourceID(
+		ipcacheTypes.ResourceKindFile,
+		cnp.ObjectMeta.Namespace,
+		cnp.ObjectMeta.Name,
+	)
```

Comment:
suggestion:

I tripped up on this, thinking you were using the CNP name instead of the filename here (which would be wrong). I think it would be easier to read if you assigned the filename into a separate variable to make sure this expression here looks the same as `deleteFromPolicyEngine`.

In addition, I think instead of using the namespace I'd use the base directory. That way, if we end up having other code in another package using `ResourceKindFile`, there is no accidental conflict because we're only using the base name here.

```suggestion
	resourceID := ipcacheTypes.NewResourceID(
		ipcacheTypes.ResourceKindFile,
		option.Config.StaticCiliumNetworkPolicyPath,
		filepath.Base(cnpFilePath),
	)
```